### PR TITLE
[SPARK-21088][ML] CrossValidator, TrainValidationSplit support collect all models when fitting: Python API

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

Add python API for collecting sub-models during CrossValidator/TrainValidationSplit fitting.

## How was this patch tested?

UT added.

### CODE REVIEW

**Comment by WeichenXu123** [comment]
Jenkins, retest this.

**Comment by WeichenXu123** [comment]
My local test passed. This test failure looks like test system issue.

**Comment by holdenk** [comment]
What happens when you run `check-license` locally? I agree it doesn't look like any of these changes would impact the license headers.

**Comment by WeichenXu123** [comment]
@holdenk Find the reason. There is an empty file in the directory. :)

**Comment by WeichenXu123** [comment]
@holdenk Thanks!

**Comment by jkbradley** [comment]
Is this still WIP or ready?

**Comment by WeichenXu123** [comment]
@jkbradley I think it is better to review #19857 (fix python model specific optimization) and merge it first  and then I rebase & update this PR. :)

**Comment by WeichenXu123** [comment]
Because of codebase changing, I will create new PR to replace this one.

**Comment by WeichenXu123** [comment]
@MrBago @yogeshg @jkbradley Updated and ready for review now!

File: python/pyspark/ml/tuning.py [hunk 1]
```diff
@@ -194,7 +195,8 @@ def _to_java_impl(self):
         return java_estimator, java_epms, java_evaluator
 
 
-class CrossValidator(Estimator, ValidatorParams, HasParallelism, MLReadable, MLWritable):
+class CrossValidator(Estimator, ValidatorParams, HasParallelism, HasCollectSubModels,
```
[the remark this answers is not in the document]
Comment (jkbradley) [reply python/pyspark/ml/tuning.py]: Let's also clarify in the doc for CrossValidatorModel.copy() that it does not copy the extra Params into the subModels.  (same for TrainValidationSplitModel)

File: python/pyspark/ml/tests.py [hunk 2]
```diff
@@ -1018,6 +1018,48 @@ def test_parallel_evaluation(self):
         cvParallelModel = cv.fit(dataset)
         self.assertEqual(cvSerialModel.avgMetrics, cvParallelModel.avgMetrics)
 
+    def test_expose_sub_models(self):
```
Comment (jkbradley) [inline python/pyspark/ml/tests.py]: Nice tests.  Can you make one addition: Test the copy() method to make sure it copies the submodels.

**Review by jkbradley — COMMENTED** [review]
Thanks for the PR!

You'll need to update _from_java and _to_java for CrossValidator and TrainValidationSplit.

Also, please update the PR description.

**Comment by jkbradley** [comment]
LGTM
Merging with master
Thanks!

**Pull request closed without merging**